manifests/params.pp

```diff
@@ -34,7 +34,24 @@
   $postrun_command     = undef
   $dns_alt_names       = []
   $use_srv_records     = false
-  $srv_domain          = $::domain
+
+  # On versions < 3.7.5 there was no defined function as such strict variables cannot be used
+  # On verisons < 4.0.0 and >= 3.5.0 the facts hash was optional disabled by default, on version < 3.5.0 the facts hash did not exist
+  # In some edge cases 'domain' fact may be not defined
+  if versioncmp($::puppetversion, '3.7.5') < 0 {
+    $srv_domain = $::domain
+  } elsif versioncmp($::puppetversion, '4.0.0') >= 0 {
```

Comment:
@domcleal the problem is that the $::domain is not even undef (it is non existent) when facter fails to evaluate. As such while yes defined helps insulate when strict vars is enabled allowing for a successful compile, when it does cause a flooding of the puppetserver logs regardless of strict var being enabled or disabled... 

On 4.x we can test for the existence of the "domain" key in the facts hash and eliminate the useless (and rather annoying) output to the logs. That is why we are testing for 4.x and not using defined on that fact.